json-schema-generator/src/main/java/io/micronaut/jsonschema/generator/CodeGenerator.java

```diff
@@ -0,0 +1,611 @@
+            objectBuilder.addSuperinterface(superInterface);
+        } else {
+            // top level class
+            addDiscriminatorAnnotations(jsonSchema, objectBuilder);
+        }
+
+        addFields(jsonSchema, objectBuilder);
+
+        if (!discriminatorProperty.isBlank()) {
+            AnnotationDef jsonTypeInfo = AnnotationDef.builder(JsonTypeInfo.class)
+                .addMember("use", JsonTypeInfo.Id.NAME)
+                .addMember("property", discriminatorProperty)
+                .build();
+            objectBuilder.addAnnotation(jsonTypeInfo);
+
+            Map<String, Schema> properties = jsonSchema.getProperties();
+            if (properties.containsKey(discriminatorProperty)) {
+                objectBuilder.addField(FieldDef.builder(discriminatorProperty)
+                    .ofType(TypeDef.STRING)
+                    .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
+                    .initializer(ExpressionDef.constant(properties.get(discriminatorProperty).getConstValue()))
+                    .build());
+            }
+        }
+        return objectBuilder.build();
+    }
+
+    private InterfaceDef buildInterface(Schema jsonSchema, String builderClassName) {
+        InterfaceDef.InterfaceDefBuilder objectBuilder = InterfaceDef.builder(builderClassName)
+            .addModifiers(Modifier.PUBLIC)
+            .addAnnotation(Serdeable.class);
+        if (jsonSchema.hasDiscriminator()) {
+            // top level interface
+            addDiscriminatorAnnotations(jsonSchema, objectBuilder);
+            if (!discriminatorProperty.isBlank()) {
+                AnnotationDef jsonTypeInfo = AnnotationDef.builder(JsonTypeInfo.class)
+                    .addMember("use", JsonTypeInfo.Id.NAME)
+                    .addMember("property", discriminatorProperty)
+                    .build();
+                objectBuilder.addAnnotation(jsonTypeInfo);
+            }
+        }
+        return objectBuilder.build();
+    }
+
+    private void addFields(Schema jsonSchema, ObjectDefBuilder builder) {
+        if (jsonSchema.hasDescription()) {
+            builder.addJavadoc(getJavadoc(jsonSchema.getDescription()));
+        }
+
+        if (jsonSchema.hasProperties()) {
+            List<String> requiredProperties = (jsonSchema.getRequired() != null) ? jsonSchema.getRequired() : new ArrayList<>();
+            jsonSchema.getProperties().forEach((key, value) -> addField(
+                builder,
+                key,
+                value,
+                requiredProperties.contains(key)
+            ));
+
+            if (jsonSchema.hasAdditionalProperties() && !jsonSchema.getAdditionalProperties().equals(Schema.FALSE)) {
+                TypeDef mapType;
+                if (jsonSchema.getAdditionalProperties().equals(Schema.TRUE)) {
+                    mapType = TypeDef.OBJECT;
+                } else {
+                    mapType = getTypeDefFromJson(jsonSchema.getAdditionalProperties());
```

Comment:
It might be a good idea to check if there is a reference in json schema at this point. If there is a reference, you can delegate to `generateFromSchema` to generate the object if needed.

Reply:
Technically, I should identify all references and their types in `saveDefinitions`. So, the `getTypeDefFromJson` would already return the saved reference's type.